include/pros/ext_adi.h

```diff
@@ -0,0 +1,641 @@
+ *
+ * \return The gyro angle in degrees.
+ */
+double ext_adi_gyro_get(ext_adi_gyro_t gyro);
+
+/**
+ * Initializes a gyroscope on the given port. If the given port has not
+ * previously been configured as a gyro, then this function starts a 1300 ms
+ * calibration period.
+ *
+ * It is highly recommended that this function be called from initialize() when
+ * the robot is stationary to ensure proper calibration.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports
+ * EADDRINUSE - The port is not configured as a gyro
+ *
+ * \param adi_port
+ *        The ADI port to initialize as a gyro (from 1-8, 'a'-'h', 'A'-'H')
+ * \param smart_port
+ *        The smart port number that the ADI Expander is in
+ * \param multiplier
+ *        A scalar value that will be multiplied by the gyro heading value
+ *        supplied by the ADI
+ *
+ * \return An adi_gyro_t object containing the given port, or PROS_ERR if the
+ * initialization failed.
+ */
+ext_adi_gyro_t ext_adi_gyro_init(uint8_t adi_port, uint8_t smart_port, double multiplier);
+
+/**
+ * Resets the gyroscope value to zero.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports
+ * EADDRINUSE - The port is not configured as a gyro
+ *
+ * \param gyro
+ *        The adi_gyro_t object for which the angle will be returned
+ *
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t ext_adi_gyro_reset(ext_adi_gyro_t gyro);
+
+/**
+ * Disables the gyro and voids the configuration on its port.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports
+ * EADDRINUSE - The port is not configured as a gyro
+ *
+ * \param gyro
+ *        The adi_gyro_t object to be shut down
+ *
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t ext_adi_gyro_shutdown(ext_adi_gyro_t gyro);
+
+
+#define SMART_PORT_BITS 16
+#define SMART_PORT_MASK ((1 << SMART_PORT_BITS) - 1)
+
+  /**
+   * Macro Description: Given a port mask, it sets the smart port and adi port to the values masked inside the int_32. 
+   */
+#define get_ports(ports, smart_port, adi_port) {\
+	uint32_t uport = (uint32_t)ports; \
+	smart_port = uport & SMART_PORT_MASK; \
+	adi_port = uport >> SMART_PORT_BITS;  }
+
+
+  /**
+   * Macro description: Given a port mask, it sets the smart port to the value masked inside the int_32. 
+   *
+   */
+#define get_smart_port(ports, smart_port) {\
+	uint32_t uport = (uint32_t)ports; \
+	smart_port = uport & SMART_PORT_MASK; \
+	}
+
+static inline uint32_t merge_adi_ports(uint8_t adi_port, uint8_t smart_port) {
+	return (adi_port << SMART_PORT_BITS) | smart_port;
+}
```

Comment:
do these really need to be in the public header? end users should not be using them